cli/internal/run/run.go

```diff
@@ -473,48 +476,136 @@ func filterSinglePackageGraphForDisplay(originalGraph *dag.AcyclicGraph) *dag.Ac
 	return graph
 }
 
-func buildTaskGraphEngine(topoGraph *dag.AcyclicGraph, pipeline fs.Pipeline, rs *runSpec) (*core.Engine, error) {
-	engine := core.NewEngine(topoGraph)
+func buildTaskGraphEngine(g *completeGraph, rs *runSpec) (*core.Engine, error) {
+	engine := core.NewEngine(&g.TopologicalGraph)
 
-	for taskName, taskDefinition := range pipeline {
-		topoDeps := make(util.Set)
+	for taskName, taskDefinition := range g.Pipeline {
 		deps := make(util.Set)
+
 		isPackageTask := util.IsPackageTask(taskName)
+
 		for _, dependency := range taskDefinition.TaskDependencies {
-			if isPackageTask && util.IsPackageTask(dependency) {
-				err := engine.AddDep(dependency, taskName)
-				if err != nil {
+			depTaskIsPackageTask := util.IsPackageTask(dependency)
+			// If both tasks are package tasks, add the relationship to the engine.
+			// Otherwise, if it's a dependency like `dependsOn: ['whatever']`
+			// add the dependency to the task itself.
+			if isPackageTask && depTaskIsPackageTask {
+				if err := engine.AddDep(dependency, taskName); err != nil {
 					return nil, err
 				}
 			} else {
 				deps.Add(dependency)
 			}
 		}
-		for _, dependency := range taskDefinition.TopologicalDependencies {
-			topoDeps.Add(dependency)
-		}
+
 		engine.AddTask(&core.Task{
-			Name:     taskName,
-			TopoDeps: topoDeps,
-			Deps:     deps,
+			Name:       taskName,
+			TopoDeps:   util.SetFromStrings(taskDefinition.TopologicalDependencies),
+			Deps:       deps,
+			Persistent: taskDefinition.Persistent,
 		})
 	}
 
-	if err := engine.Prepare(&core.EngineBuildingOptions{
+	engineOptions := &core.EngineBuildingOptions{
 		Packages:  rs.FilteredPkgs.UnsafeListOfStrings(),
 		TaskNames: rs.Targets,
 		TasksOnly: rs.Opts.runOpts.only,
-	}); err != nil {
+	}
+
+	if err := engine.Prepare(engineOptions); err != nil {
 		return nil, err
 	}
 
+	// Check for cycles in the DAG.
 	if err := util.ValidateGraph(engine.TaskGraph); err != nil {
 		return nil, fmt.Errorf("Invalid task dependency graph:\n%v", err)
 	}
 
+	errorThing := validatePersistentDependencies(engine, g)
+	if errorThing.PackageName != "" {
```

Comment:
not a fan of checking an arbitrary field in the struct to see if it's populated, but I couldn't figure out how to return `nil` or `ErrorThingy` from the method (Super Go and static typing beginner 👋🏾 )

Reply:
I think what you want to do is implement [`Error`](https://gobyexample.com/errors) for `ErrorThingy` and change the return type of `validatePersistentDependencies` to `error` which will allow you to return `nil`.